// @gorhom/animated-tabbar/lib/commonjs/presets/material/item/MaterialTabBarItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNativeReanimated = _interopRequireDefault(require("react-native-reanimated"));

var _reactNativeRedash = require("react-native-redash");

var _lodash = _interopRequireDefault(require("lodash.isequal"));

var _hooks = require("../../../hooks");

var _utilities = require("../../../utilities");

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// @ts-ignore 😞
const {
  divide,
  Extrapolate
} = _reactNativeReanimated.default;

const MaterialTabBarItemComponent = props => {
  // props
  const {
    animatedFocus,
    animation,
    inactiveOpacity,
    inactiveScale,
    label,
    labelStyle: labelStyleOverride,
    icon,
    iconSize,
    spacing
  } = props; //#region extract props

  const {
    innerVerticalSpace,
    innerHorizontalSpace,
    outerVerticalSpace,
    outerHorizontalSpace
  } = spacing; //#endregion
  //#region variables

  const labelHeight = (0, _reactNativeRedash.useValue)(0); //#endregion
  //#region styles

  const outerContainerStyle = (0, _react.useMemo)(() => [_styles.styles.outerContainer, {
    paddingHorizontal: outerHorizontalSpace,
    paddingVertical: outerVerticalSpace
  }], [outerHorizontalSpace, outerVerticalSpace]);
  const innerContainerStyle = (0, _react.useMemo)(() => [_styles.styles.innerContainer, {
    paddingHorizontal: innerHorizontalSpace,
    paddingVertical: innerVerticalSpace,
    opacity: (0, _utilities.interpolate)(animatedFocus, {
      inputRange: [0, 1],
      outputRange: [inactiveOpacity, 1],
      extrapolate: Extrapolate.CLAMP
    }),
    transform: (0, _reactNativeRedash.transformOrigin)({
      x: 0,
      y: 0
    }, {
      scale: (0, _utilities.interpolate)(animatedFocus, {
        inputRange: [0, 1],
        outputRange: [inactiveScale, 1],
        extrapolate: Extrapolate.CLAMP
      })
    })
  }], [innerHorizontalSpace, innerVerticalSpace, animatedFocus, inactiveOpacity, inactiveScale]);
  const iconContainerStyle = [_styles.styles.iconContainer, {
    minHeight: iconSize,
    minWidth: iconSize,
    ...(animation === 'iconWithLabelOnFocus' ? {
      transform: (0, _reactNativeRedash.transformOrigin)({
        x: 0,
        y: 0
      }, {
        translateY: (0, _utilities.interpolate)(animatedFocus, {
          inputRange: [0, 1],
          outputRange: [divide(labelHeight, 2), 0],
          extrapolate: Extrapolate.CLAMP
        })
      })
    } : {})
  }];
  const labelStyle = (0, _react.useMemo)(() => [_styles.styles.label, labelStyleOverride, {
    marginTop: innerVerticalSpace / 2,
    opacity: animation === 'iconWithLabelOnFocus' ? animatedFocus : 1
  }], [labelStyleOverride, innerVerticalSpace, animation, animatedFocus]); //#endregion
  //#region callbacks

  const handleLabelLayout = (0, _hooks.useStableCallback)(({
    nativeEvent: {
      layout: {
        height
      }
    }
  }) => requestAnimationFrame(() => {
    labelHeight.setValue(height);
  })); //#endregion
  // render

  const renderIcon = (0, _react.useCallback)(() => {
    const IconComponent = icon.component;
    return typeof IconComponent === 'function' ? IconComponent({
      animatedFocus,
      color: icon.color,
      size: iconSize
    }) : /*#__PURE__*/_react.default.createElement(IconComponent, {
      animatedFocus: animatedFocus,
      color: icon.color,
      size: iconSize
    });
  }, [icon, iconSize, animatedFocus]);
  const renderLabel = (0, _react.useCallback)(() => {
    return animation !== 'iconOnly' ? /*#__PURE__*/_react.default.createElement(_reactNativeReanimated.default.Text, {
      onLayout: handleLabelLayout,
      style: labelStyle
    }, label) : null;
  }, [label, animation, labelStyle, handleLabelLayout]);
  return /*#__PURE__*/_react.default.createElement(_reactNativeReanimated.default.View, {
    style: outerContainerStyle
  }, /*#__PURE__*/_react.default.createElement(_reactNativeReanimated.default.View, {
    style: innerContainerStyle
  }, /*#__PURE__*/_react.default.createElement(_reactNativeReanimated.default.View, {
    style: iconContainerStyle
  }, renderIcon()), renderLabel()));
};

const MaterialTabBarItem = /*#__PURE__*/(0, _react.memo)(MaterialTabBarItemComponent, _lodash.default);
var _default = MaterialTabBarItem;
exports.default = _default;
